File: challenge-1/go/shop_discount.go
/*
Sistema di gestione sconti per uno shop online.

Il codice funziona, ma è pieno di magic numbers e magic strings.
Estrai tutte le costanti e dai loro nomi significativi!
*/
package main

import (
	"fmt"
	"math"
	"strings"
)

// Item rappresenta un prodotto con prezzo e quantità
type Item struct {
	P float64
	Q int
}

// Order contiene i dati di un ordine
type Order struct {
	Subtotal        float64
	Tax             float64
	Total           float64
	DiscountApplied bool
}

/*
CalculateDiscount calcola il prezzo finale con sconto.

Parametri:
  - price: prezzo unitario del prodotto
  - customerType: tipo di cliente ("basic", "premium", "vip")
  - quantity: quantità di prodotti acquistati

Ritorna il prezzo finale dopo applicazione di tutti gli sconti
*/
func CalculateDiscount(price float64, customerType string, quantity int) float64 {
	// Sconto per clienti premium
	if customerType == "premium" {
		price = price * 0.85
	}

	// Sconto per clienti VIP
	if customerType == "vip" {
		price = price * 0.75
	}

	// Sconto per quantità
	if quantity >= 10 {
		price = price * 0.9
	} else if quantity >= 5 {
		price = price * 0.95
	}

	// Sconto minimo garantito per ordini grandi
	if price > 100 {
		price = price * 0.98
	}

	return math.Round(price*100) / 100
}

/*
CalculateTotalOrder calcola il totale di un ordine applicando sconti per cliente e tasse.

Parametri:
  - products: slice di Product
  - customerType: tipo di cliente ("basic", "premium", "vip")

Ritorna una struct OrderData con subtotale, tasse e totale finale
*/
func CalculateTotalOrder(items []Item, customerType string) Order {
	s := 0.0

	for _, i := range items {
		discountedPrice := CalculateDiscount(i.P, customerType, i.Q)
		s = s + (discountedPrice * float64(i.Q))
	}

	// Tassa sul valore
	t := s * 0.22

	tot := s + t

	// Sconto finale se l'ordine è superiore a 500 euro
	if tot > 500 {
		tot = tot * 0.95
	}

	return Order{
		Subtotal:        math.Round(s*100) / 100,
		Tax:             math.Round(t*100) / 100,
		Total:           math.Round(tot*100) / 100,
		DiscountApplied: tot != s+t,
	}
}

/*
GetCustomerTierDescription ritorna una descrizione del livello cliente.
*/
func GetCustomerTierDescription(cType string) string {
	switch cType {
	case "basic":
		return "Cliente Base"
	case "premium":
		return "Cliente Premium (15% sconto base)"
	case "vip":
		return "Cliente VIP (25% sconto base)"
	default:
		return "Tipo cliente sconosciuto"
	}
}

/*
FormatOrderReceipt formatta uno scontrino leggibile.
*/
func FormatOrderReceipt(orderData Order, customerType string) string {
	receipt := ""
	receipt += strings.Repeat("=", 40) + "\n"
	receipt += "SCONTRINO ORDINE\n"
	receipt += strings.Repeat("=", 40) + "\n"
	receipt += fmt.Sprintf("Tipo cliente: %s\n", GetCustomerTierDescription(customerType))
	receipt += strings.Repeat("-", 40) + "\n"
	receipt += fmt.Sprintf("Subtotale: € %.2f\n", orderData.Subtotal)
	receipt += fmt.Sprintf("Tasse (22%%): € %.2f\n", orderData.Tax)

	if orderData.DiscountApplied {
		receipt += "SCONTO PER ORDINE GRANDE: -5%\n"
	}

	receipt += strings.Repeat("-", 40) + "\n"
	receipt += fmt.Sprintf("TOTALE: € %.2f\n", orderData.Total)
	receipt += strings.Repeat("=", 40) + "\n"

	return receipt
}

// Esempio di utilizzo
func main() {
	products := []Item{
		{P: 50, Q: 2}, // 2x prodotto da €50
		{P: 30, Q: 8}, // 8x prodotto da €30
	}

	order := CalculateTotalOrder(products, "premium")
	fmt.Print(FormatOrderReceipt(order, "premium"))

	fmt.Println("\nAltro esempio:")
	products2 := []Item{
		{P: 100, Q: 3},
		{P: 50, Q: 5},
	}
	order2 := CalculateTotalOrder(products2, "vip")
	fmt.Print(FormatOrderReceipt(order2, "vip"))
}

File: challenge-1/java/ShopDiscount.java
import java.util.List;

/**
 * Sistema di gestione sconti per uno shop online.
 * <p>
 * Il codice funziona, ma è pieno di magic numbers e magic strings.
 * Estrai tutte le costanti e dai loro nomi significativi!
 */
public class ShopDiscount {

    public static final double STANDARD_DISCOUNT = 0.85;
    public static final double PREMIUM_DISCOUNT = 0.85;
    public static final double VIP_DISCOUNT = 0.75;
    public static final double QUANTITY_DISCOUNT_5 = 0.95;
    public static final double QUANTITY_DISCOUNT_10 = 0.9;
    public static final double LARGE_ORDER_DISCOUNT_THRESHOLD = 100.0;
    public static final double LARGE_ORDER_DISCOUNT = 0.98;
    public static final double TAX_RATE = 0.22;
    public static final double TOTAL_ORDER_DISCOUNT_THRESHOLD = 500.0;
    public static final double TOTAL_ORDER_DISCOUNT = 0.95;

    public record Item(
            double price,
            int quantity
    ) {
    }

    public record Order(
            double subtotal,
            double tax,
            double total,
            boolean discountApplied
    ) {
    }

    /**
     * Calcola il prezzo finale con sconto.
     *
     * @param price        prezzo unitario del prodotto
     * @param customerType tipo di cliente ("basic", "premium", "vip")
     * @param quantity     quantità di prodotti acquistati
     * @return prezzo finale dopo applicazione di tutti gli sconti
     */
    public static double calculateDiscount(double price, String customerType, int quantity) {
        // Sconto per clienti premium
        if (customerType.equals("premium")) {
            price = price * PREMIUM_DISCOUNT;
        }

        // Sconto per clienti VIP
        if (customerType.equals("vip")) {
            price = price * VIP_DISCOUNT;
        }

        // Sconto per quantità
        if (quantity >= 10) {
            price = price * QUANTITY_DISCOUNT_10;
        } else if (quantity >= 5) {
            price = price * QUANTITY_DISCOUNT_5;
        }

        // Sconto minimo garantito per ordini grandi
        if (price > 100) {
            price = price * LARGE_ORDER_DISCOUNT;
        }

        return Math.round(price * 100.0) / 100.0;
    }

    /**
     * Calcola il totale di un ordine applicando sconti per cliente e tasse.
     *
     * @param products     lista di coppie [p, q]
     * @param customerType tipo di cliente ("basic", "premium", "vip")
     * @return mappa con subtotale, tasse e totale finale
     */
    public static Order calculateTotalOrder(List<Item> products, String customerType) {
        double subtotal = 0;

        for (Item i : products) {
            double discount = calculateDiscount(i.price(), customerType, i.quantity());
            subtotal = subtotal + (discount * i.quantity());
        }

        // Tassa sul valore
        double tax = subtotal * 0.22;

        double tot = subtotal + tax;

        // Sconto finale se l'ordine è superiore a 500 euro
        if (tot > 500) {
            tot = tot * TOTAL_ORDER_DISCOUNT;
        }

        Order result = new Order(
                Math.round(subtotal * 100.0) / 100.0,
                Math.round(tax * 100.0) / 100.0,
                Math.round(tot * 100.0) / 100.0,
                tot != subtotal + tax
        );
        return result;
    }

    /**
     * Ritorna una descrizione del livello cliente.
     */
    public static String getCustomerTierDescription(String customerType) {
        return switch (customerType) {
            case "basic" -> "Cliente Base";
            case "premium" -> "Cliente Premium (15% sconto base)";
            case "vip" -> "Cliente VIP (25% sconto base)";
            default -> "Tipo cliente sconosciuto";
        };
    }

    /**
     * Formatta uno scontrino leggibile.
     */
    public static String formatOrderReceipt(Order orderData, String customerType) {
        StringBuilder receipt = new StringBuilder();
        receipt.append("=".repeat(40)).append("\n");
        receipt.append("SCONTRINO ORDINE\n");
        receipt.append("=".repeat(40)).append("\n");
        receipt.append("Tipo cliente: ").append(getCustomerTierDescription(customerType)).append("\n");
        receipt.append("-".repeat(40)).append("\n");
        receipt.append("Subtotale: € ").append(orderData.subtotal()).append("\n");
        receipt.append("Tasse (22%): € ").append(orderData.tax()).append("\n");

        if (orderData.discountApplied()) {
            receipt.append("SCONTO PER ORDINE GRANDE: -5%\n");
        }

        receipt.append("-".repeat(40)).append("\n");
        receipt.append("TOTALE: € ").append(orderData.total()).append("\n");
        receipt.append("=".repeat(40)).append("\n");

        return receipt.toString();
    }

    // Esempio di utilizzo
    public static void main(String[] args) {
        List<Item> products = List.of(
                new Item(50, 2),    // 2x prodotto da €50
                new Item(30, 8)     // 8x prodotto da €30
        );

        Order order = calculateTotalOrder(products, "premium");
        System.out.println(formatOrderReceipt(order, "premium"));

        System.out.println("\nAltro esempio:");
        List<Item> products2 = List.of(
                new Item(100, 3),
                new Item(50, 5)
        );
        Order order2 = calculateTotalOrder(products2, "vip");
        System.out.println(formatOrderReceipt(order2, "vip"));
    }
}
